clippy_lints/src/transmute/missing_transmute_annotations.rs

```diff
@@ -0,0 +1,53 @@
+use rustc_errors::Applicability;
+use rustc_hir::{GenericArg, HirId, Node, Path, TyKind};
+use rustc_lint::LateContext;
+use rustc_middle::lint::in_external_macro;
+use rustc_middle::ty::Ty;
+
+use clippy_utils::diagnostics::span_lint_and_sugg;
+
+use crate::transmute::MISSING_TRANSMUTE_ANNOTATIONS;
+
+pub(super) fn check<'tcx>(
+    cx: &LateContext<'tcx>,
+    path: &Path<'tcx>,
+    from_ty: Ty<'tcx>,
+    to_ty: Ty<'tcx>,
+    expr_hir_id: HirId,
+) -> bool {
+    let last = path.segments.last().unwrap();
+    if in_external_macro(cx.tcx.sess, last.ident.span) {
+        // If it comes from a non-local macro, we ignore it.
+        return false;
+    }
+    let args = last.args;
+    let missing_generic = match args {
+        Some(args) if !args.args.is_empty() => args.args.iter().any(|arg| match arg {
+            GenericArg::Infer(_) => true,
+            GenericArg::Type(ty) => matches!(ty.kind, TyKind::Infer),
+            _ => false,
+        }),
+        _ => true,
+    };
+    if !missing_generic {
+        return false;
+    }
+    // If it's being set as a local variable value...
```

Comment:
We should include closures here imo, if they have type annotations. This could include the parameter if it's specified as well, but not something like:
```rust
// i64 is known, but [i32; 2] is never specified
|x: i32, y: i32| -> i64 unsafe { transmute::<[i32; 2], _>([x, y]) };
```

Reply:
I don't think we should special case closures either. It adds a lot of complexity to detect such cases for a very marginal gain imo.

Reply:
Not really. The return type part is easy, and the parameter is too - Get the `Res` from the `QPath`, the `HirId` of the local, then the parent's parent *should* include the types in `FnDecl`.

I wouldn't block it on this so if it's not done I can add it after, eventually.

Reply:
Not in this sense. I meant, for the same reason we want to enforce having type annotations when `transmute` is used as return expr, we want to enforce for it for closures as well.

Reply:
I don't agree with that as closures are usually private and not public. It's more akin to `let` than functions. I don't think you can fold a closure either in IDEs.

Reply:
I don't see why closures being private or public changes anything. If you want to bring this back to debate, let's continue on zulip. ;)

Reply:
I'll bring up my current points on zulip sometime.

Reply:
:+1: There is an open thread already, don't hesitate to comment there directly. ;)